src/components/Provider.js

```diff
@@ -20,39 +22,83 @@ function warnAboutReceivingStore() {
 }
 
 export function createProvider(storeKey = 'store') {
-    const subscriptionKey = `${storeKey}Subscription`
 
     class Provider extends Component {
-        getChildContext() {
-          return { [storeKey]: this[storeKey], [subscriptionKey]: null }
+
+        constructor(props) {
+          super(props)
+
+            const {store} = props;
+
+            this.state = {
+                storeState : store.getState(),
+                store,
+            };
+        }
+
+        componentDidMount() {
+            this.subscribe();
+        }
+
+        componentWillUnmount() {
+          if(this.unsubscribe) {
+            this.unsubscribe()
+            this._isMounted = false;
+          }
         }
 
-        constructor(props, context) {
-          super(props, context)
-          this[storeKey] = props.store;
+        subscribe() {
+          const {store} = this.props;
+
+          this._isMounted = true;
+
+          this.unsubscribe = store.subscribe( () => {
+            const newStoreState = store.getState();
+
+            if(!this._isMounted) {
+              return;
+            }
+
+            this.setState(providerState => {
+              // If the value is the same, skip the unnecessary state update.
+              if(providerState.storeState === newStoreState) {
+                return null;
+              }
+
+              return {storeState : newStoreState};
+            })
+          });
+
+          // Actions might have been dispatched between render and mount - handle those
+          const postMountStoreState = store.getState();
+          if(postMountStoreState !== this.state.storeState) {
+            this.setState({storeState : postMountStoreState});
+          }
         }
 
         render() {
-          return Children.only(this.props.children)
+            return (
+                <ReactReduxContext.Provider value={this.state}>
+                    {Children.only(this.props.children)}
+                </ReactReduxContext.Provider>
+            );
         }
     }
 
     if (process.env.NODE_ENV !== 'production') {
-      Provider.prototype.componentDidUpdate = function () {
-        if (this[storeKey] !== this.props.store) {
+      Provider.getDerivedStateFromProps = function (props, state) {
+        if (state.store !== props.store) {
           warnAboutReceivingStore()
```

Comment:
this is not required with the new context any longer, we can change the context value and children are updated

Reply:
But, the selector initialization and binding of action creators happens when the connected children are instantiated - we don't try to re-bind those if the store is swapped out.

Reply:
We can rebind in cDU, 1 line change

Reply:
Not something I'm particularly worried about right now.  There may be _someone_ out there who has a need to swap out their store, but it's not something we've ever supported and I don't see a need to try to support it for the time being.